File: GuideObjC/GuideObjC/ViewControllers/NewPostController/NewPostController.m
//
//  NewPostController.m
//  GuideObjC
//
//  Created by LuzanovRoman on 12.12.16.
//  Copyright © 2016 eigengraph. All rights reserved.
//

#import "NewPostController.h"
#import "ProgressController.h"

@interface NewPostController ()
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *textViewHeight;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *imageHeight;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *imageWidth;
@property (weak, nonatomic) IBOutlet UIBarButtonItem *sendButton;
@property (weak, nonatomic) IBOutlet UILabel *placeholder;
@property (weak, nonatomic) IBOutlet UILabel *errorLabel;
@property (weak, nonatomic) IBOutlet UITextView *textView;
@property (weak, nonatomic) IBOutlet UIButton *imageButton;
@property (retain, nonatomic) UIImage *postImage;
@property (retain, nonatomic) NSString *currentUserId;
@end

@implementation NewPostController
@synthesize postImage, currentUserId;

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self.graph userObjectWithCompletion:^(NSObject * object, NSError * error) {
        if ([object isKindOfClass:[EGFUser class]]) {
            currentUserId = ((EGFUser *)object).id;
        }
    }];
}

- (void)showError:(NSError *)error {
    if (error) {
        _errorLabel.text = error.localizedFailureReason;
    }
    else {
        _errorLabel.text = @"An unknown error";
    }
}

- (IBAction)createPost:(id)sender {
    _errorLabel.text = nil;
    
    if (currentUserId && postImage) {
        NSData * imageData = UIImageJPEGRepresentation(postImage, 0.75);
        NSString * desc = _textView.text;
        
        [ProgressController show];
        [self.graph uploadImageWithData:imageData title:@"Photo" mimeType:@"image/jpeg" kind:@"image" completion:^(NSObject * object, NSError * error) {
            if ([object isKindOfClass:[EGFFile class]]) {
                NSDictionary * parameters = @{@"image":((EGFFile *)object).id, @"desc":desc, @"object_type":@"post"};
                [self.graph createObjectWithParameters:parameters forSource:currentUserId onEdge:@"posts" completion:^(NSObject * object, NSError * error) {
                    if ([object isKindOfClass:[EGFPost class]]) {
                        [self.navigationController popViewControllerAnimated:true];
                    }
                    else {
                        [self showError:error];
                    }
                    [ProgressController hide];
                }];
            }
            else {
                [self showError:error];
                [ProgressController hide];
            }
        }];
    }
}

- (IBAction)imageButtonDidTouchUp:(id)sender {
    UIAlertController * controller = [UIAlertController alertControllerWithTitle:nil message:nil preferredStyle:UIAlertControllerStyleActionSheet];
    [controller addAction:[UIAlertAction actionWithTitle:@"Take Photo" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [self getImageWithSource:UIImagePickerControllerSourceTypeCamera];
    }]];
    [controller addAction:[UIAlertAction actionWithTitle:@"From Camera Roll" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [self getImageWithSource:UIImagePickerControllerSourceTypePhotoLibrary];
    }]];
    [controller addAction:[UIAlertAction actionWithTitle:@"Cancel" style:UIAlertActionStyleCancel handler:nil]];
    [self presentViewController:controller animated:true completion:nil];
}

- (void)getImageWithSource:(UIImagePickerControllerSourceType)sourceType {
    UIImagePickerController * controller = [[UIImagePickerController alloc] init];
    controller.sourceType = sourceType;
    controller.delegate = self;
    [self presentViewController:controller animated:true completion:nil];
}

- (void)updateSendButton {
    _sendButton.enabled = postImage != nil && _textView.text.length > 0;
}

// MARK:- UITextViewDelegate
- (void)textViewDidChange:(UITextView *)textView {

    CGSize size = [_textView sizeThatFits:CGSizeMake(textView.frame.size.width, 1000)];
    CGFloat newHeight = MIN(100, MAX(50, size.height));
    
    if (newHeight != _textViewHeight.constant) {
        _textViewHeight.constant = newHeight;
    }
    _placeholder.hidden = textView.text.length > 0;
    [self updateSendButton];
}

- (BOOL)textView:(UITextView *)textView shouldChangeTextInRange:(NSRange)range replacementText:(NSString *)text {
    if ([text isEqual:@"\n"]) {
        [textView resignFirstResponder];
        return false;
    }
    return true;
}

// MARK:- UIImagePickerControllerDelegate
- (void)imagePickerController:(UIImagePickerController *)picker didFinishPickingMediaWithInfo:(NSDictionary<NSString *,id> *)info {
    [picker dismissViewControllerAnimated:true completion:^{
        UIImage * image = info[UIImagePickerControllerOriginalImage];
        
        if (image) {
            CGFloat maxWidth = self.view.frame.size.width - 40;
            CGFloat imageRatio = image.size.width / image.size.height;
            
            _imageWidth.constant = MIN(maxWidth, image.size.width);
            _imageHeight.constant = _imageWidth.constant / imageRatio;
            [_imageButton setImage:image forState:UIControlStateNormal];
            [_imageButton setTitle:nil forState:UIControlStateNormal];

            CGFloat imageMaxSize = 800;
            
            if (image.size.width > imageMaxSize || image.size.height > imageMaxSize) {
                CGFloat w = MIN(imageMaxSize, image.size.width);
                CGFloat h = w / imageRatio;
                
                CGImageRef cropImageRef = CGImageCreateWithImageInRect(image.CGImage, CGRectMake(0, 0, w, h));
                self.postImage = [UIImage imageWithCGImage:cropImageRef];
                CGImageRelease(cropImageRef);
            }
            else {
                self.postImage = image;
            }
            [self updateSendButton];
        }
    }];
}

- (void)imagePickerControllerDidCancel:(UIImagePickerController *)picker {
    [picker dismissViewControllerAnimated:true completion:nil];
}
@end

File: GuideSwift/GuideSwift/ViewControllers/NewPostController/NewPostController.swift
//
//  NewPostController.swift
//  GuideSwift
//
//  Created by LuzanovRoman on 12.12.16.
//  Copyright © 2016 eigengraph. All rights reserved.
//

import UIKit

class NewPostController: BaseController, UITextViewDelegate, UIActionSheetDelegate, UIImagePickerControllerDelegate, UINavigationControllerDelegate {
    @IBOutlet weak var textViewHeight: NSLayoutConstraint!
    @IBOutlet weak var imageHeight: NSLayoutConstraint!
    @IBOutlet weak var imageWidth: NSLayoutConstraint!
    @IBOutlet weak var sendButton: UIBarButtonItem!
    @IBOutlet weak var placeholder: UILabel!
    @IBOutlet weak var errorLabel: UILabel!
    @IBOutlet weak var textView: UITextView!
    @IBOutlet weak var imageButton: UIButton!
    fileprivate var postImage: UIImage?
    fileprivate var currentUserId: String?

    override func viewDidLoad() {
        super.viewDidLoad()

        Graph.userObject { (object, _) in
            self.currentUserId = (object as? EGFUser)?.id
        }
    }

    @IBAction func createPost(_ sender: AnyObject) {
        errorLabel.text = nil

        guard let userId = currentUserId, let image = postImage, let imageData = UIImageJPEGRepresentation(image, 0.75),
            let desc = textView.text, !desc.isEmpty else { return }

        ProgressController.show()
        Graph.uploadImage(withData: imageData, title: "Photo", mimeType: "image/jpeg", kind: "image") { (object, error) in
            if let file = object as? EGFFile, let fileId = file.id {
                let parameters = ["image": fileId, "desc": desc, "object_type": "post"]
                Graph.createObject(withParameters: parameters, forSource: userId, onEdge: "posts") { (object, error) in
                    if let _ = object as? EGFPost {
                        _ = self.navigationController?.popViewController(animated: true)
                    } else {
                        self.show(error: error)
                    }
                    ProgressController.hide()
                }
            } else {
                self.show(error: error)
                ProgressController.hide()
            }
        }
    }

    @IBAction func imageButtonDidTouchUp(_ sender: AnyObject) {
        let controller = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
        controller.addAction(UIAlertAction(title: "Take Photo", style: .default) { (_) in
            self.getImage(withSource: .camera)
        })
        controller.addAction(UIAlertAction(title: "From Camera Roll", style: .default) { (_) in
            self.getImage(withSource: .photoLibrary)
        })
        controller.addAction(UIAlertAction(title: "Cancel", style: .cancel, handler: nil))
        present(controller, animated: true, completion: nil)
    }

    fileprivate func getImage(withSource sourceType: UIImagePickerControllerSourceType) {
        let controller = UIImagePickerController()
        controller.sourceType = sourceType
        controller.delegate = self
        present(controller, animated: true, completion: nil)
    }

    fileprivate func show(error: NSError?) {
        if let err = error {
            self.errorLabel.text = err.localizedFailureReason
        } else {
            self.errorLabel.text = "An unknown error"
        }
    }

    fileprivate func updateSendButton() {
        sendButton.isEnabled = postImage != nil && !textView.text.isEmpty
    }

    // MARK: - UITextViewDelegate
    func textViewDidChange(_ textView: UITextView) {
        let size = textView.sizeThatFits(CGSize(width: textView.frame.size.width, height: 1000))
        let newHeight = min(100, max(50, size.height))

        if newHeight != textViewHeight.constant {
            textViewHeight.constant = newHeight
            view.layoutIfNeeded()
            textView.setContentOffset(CGPoint(x: 0, y: 0), animated: false)
        }
        placeholder.isHidden = !textView.text.isEmpty
        updateSendButton()
    }

    func textView(_ textView: UITextView, shouldChangeTextIn range: NSRange, replacementText text: String) -> Bool {
        if text == "\n" {
            textView.resignFirstResponder()
            return false
        }
        return true
    }

    // MARK: - UIImagePickerControllerDelegate
    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [String : Any]) {
        picker.dismiss(animated: true) { () -> Void in
            guard let image = info[UIImagePickerControllerOriginalImage] as? UIImage else { return }

            let maxWidth = self.view.frame.size.width - 40
            let imageRatio = image.size.width / image.size.height
            self.imageWidth.constant = min(maxWidth, image.size.width)
            self.imageHeight.constant = self.imageWidth.constant / imageRatio
            self.imageButton.setImage(image, for: .normal)
            self.imageButton.setTitle(nil, for: .normal)

            let imageMaxSize: CGFloat = 800

            if image.size.width > imageMaxSize || image.size.height > imageMaxSize {
                let maxValue = max(image.size.width, image.size.height)
                let scale = maxValue / imageMaxSize
                let size = CGSize(width: image.size.width / scale, height: image.size.height / scale)
                self.postImage = UIImage.image(withImage: image, scaledToSize: size)
            } else {
                self.postImage = image
            }
            self.updateSendButton()
        }
    }

    func imagePickerControllerDidCancel(_ picker: UIImagePickerController) {
        picker.dismiss(animated: true, completion: nil)
    }
}
